### ISSUE TITLE
Add definable property for typed values on a thing

### ISSUE THREAD

**Issue by joente** [thingsdb/ThingsDB#210]
**Problem**
It is useful to use a thing as a lookup for different reasons. This is already possible in ThingsDB but to specify such lookup in a `Type` you have to use `thing` as the most strict option. When doing this, the lookup is allowed to have any property to be assigned. Instead, we would like to tell ThingsDB that only properties of a certain type are allowed.

**Solution**
I would like ThingsDB to allow the definable property `"thing<..>"`.

This could be used like:

* `"thing<int>"`: Every value in the *thing* must be of type `int`.
* `"thing<int?>"`: Every value in the *thing* must be of type `int` or `nil`.
* `"thing<Foo>"`: Every value in the *thong* must by of type `Foo`.

**Context**
To implement the above, the internal `ti_thing_t` must be changed to add a reference to the *field* with the condition. Therefore, this is a huge change and should be implemented in version 1.x.

**Comment by joente** [comment]
**How to store the restriction info?**

- When a *new* thing is being packed, function `ti_thing_to_pk(..)` should pack the thing different
```c
    if (options == TI_VAL_PACK_TASK)
    {
        if (ti_thing_is_new(thing))
        {
            ti_thing_unmark_new(thing);
            // Here we should check and call a different function for restricted things
            return ti_thing_is_object(thing)  
                    ? ti_thing__to_pk(thing, vp, options)
```

The new function should use a special char like `(`.

- If a thing *with* an Id is restricted, a new *task* should be added to restrict an existing thing.
- Full storage should be able to handle the store of the restriction.

**Comment by joente** [comment]
Implemented in **v1.1.1-rc0**
